test/performance/odp_dma_perf.c

```diff
@@ -9,266 +9,511 @@
+static test_api_t test_api;
+
+static void terminate(int signal ODP_UNUSED)
 {
-	memset(config, 0, sizeof(*config));
-	config->num_in_seg = 1;
-	config->seg_size = DEFAULT_SEG_SIZE;
-	config->num_rounds = ROUNDS;
-	config->wait_ns = DEFAULT_WAIT_NS;
-	config->compl_modes.compl_mask = ODP_DMA_COMPL_SYNC;
+	odp_atomic_store_u32(&is_running, 0U);
 }
 
-static void parse_completion_modes(test_config_t *config, const char *optarg)
+static void init_config(prog_config_t *config)
 {
-	char *tmp_str = strdup(optarg);
-	char *tmp;
-	int mode;
-	uint32_t i = 0U;
-
-	config->compl_modes.num_modes = 0;
-
-	if (tmp_str == NULL)
-		return;
+	sd_t *sd;
+	trs_info_t *info;
+	stats_t *stats;
 
-	tmp = strtok(tmp_str, COMPL_DELIMITER);
+	memset(config, 0, sizeof(*config));
+	config->compl_mode_mask |= ODP_DMA_COMPL_SYNC;
+	config->num_in_segs = DEF_SEG_CNT;
+	config->num_out_segs = DEF_SEG_CNT;
+	config->src_seg_len = DEF_LEN;
+	config->num_inflight = DEF_INFLIGHT;
+	config->time_sec = DEF_TIME;
+	config->num_workers = DEF_WORKERS;
+	config->trs_type = DEF_TRS_TYPE;
+	config->seg_type = DEF_SEG_TYPE;
+	config->compl_mode = DEF_MODE;
+	config->policy = DEF_POLICY;
+
+	for (uint32_t i = 0U; i < MAX_WORKERS; ++i) {
+		sd = &sds[i];
+		stats = &config->thread_config[i].stats;
+		memset(sd, 0, sizeof(*sd));
+
+		for (uint32_t i = 0U; i < MAX_SEGS; ++i) {
+			info = &sd->dma.infos[i];
+			info->compl_param.transfer_id = ODP_DMA_TRANSFER_ID_INVALID;
+			info->compl_param.event = ODP_EVENT_INVALID;
+			info->compl_param.queue = ODP_QUEUE_INVALID;
+			odp_ticketlock_init(&info->lock);
+			sd->seg.src_pkt[i] = ODP_PACKET_INVALID;
+			sd->seg.dst_pkt[i] = ODP_PACKET_INVALID;
+		}
 
-	while (tmp) {
-		mode = atoi(tmp);
-		config->compl_modes.modes[i] = mode;
-		config->compl_modes.compl_mask |= compl_mode_map[mode];
-		++i;
-		++config->compl_modes.num_modes;
-		tmp = strtok(NULL, COMPL_DELIMITER);
+		sd->dma.handle = ODP_DMA_INVALID;
+		sd->dma.pool = ODP_POOL_INVALID;
+		sd->dma.compl_q = ODP_QUEUE_INVALID;
+		sd->seg.src_pool = ODP_POOL_INVALID;
+		sd->seg.dst_pool = ODP_POOL_INVALID;
+		sd->seg.src_shm = ODP_SHM_INVALID;
+		sd->seg.dst_shm = ODP_SHM_INVALID;
+		sd->grp = ODP_SCHED_GROUP_INVALID;
+		stats->min_trs_tm = UINT64_MAX;
+		stats->min_start_cc = UINT64_MAX;
+		stats->min_wait_cc = UINT64_MAX;
+		stats->min_trs_cc = UINT64_MAX;
 	}
-
-	free(tmp_str);
 }
 
 static void print_usage(void)
 {
 	printf("\n"
-	       "DMA performance test. Transfers a set of source segments to a single destination\n"
-	       "segment.\n"
+	       "DMA performance test. Load DMA subsystem from several workers.\n"
 	       "\n"
 	       "Examples:\n"
-	       "    odp_dma_perf\n"
-	       "    odp_dma_perf -t 0 -g 1 -i 2\n"
-	       "    odp_dma_perf -t 1 -g 1 -i 4 -m 0,0,0,0\n"
-	       "    odp_dma_perf -t 1 -g 1 -i 7 -m 0,0,0,0,0,0,1 -T 1 -r 1000 -s 2048\n"
+	       "    " PROG_NAME " -s 10240\n"
+	       "    " PROG_NAME " -t 0 -i 1 -o 1 -s 16777215 -S 1 -f 128 -T 10\n"
```

Comment:
Neither of these examples seem to work on linux-generic (applied 912bdf714 manually). Also, it would be more user-friendly if the default values for command line options would be "safer", so simply running the application without any options could work.

Reply:
Adjusted the default values in v2, now works also without any explicit parameters. The examples are just examples, they do work on some implementations.

Reply:
Yes, they are just examples, but often when testing a new application for the first time users simply copy-paste the commands from help. So it's good to use relatively basic options in the example commands to increase the probability that they work on different platforms.

Reply:
Slightly less bad guesses in v3.